full-node/sov-prover-storage-manager/src/snapshot_manager.rs

```diff
@@ -55,9 +61,140 @@ impl SnapshotManager {
+
+            snapshot_iterators.push(parent_snapshot.iter::<S>());
+
+            snapshot_id = *parent_snapshot_id;
+        }
+
+        snapshot_iterators.reverse();
+        let db_iter = self.db.raw_iter::<S>()?;
+
+        Ok(SnapshotManagerIter::new(db_iter, snapshot_iterators))
+    }
+}
+
+/// [`Iterator`] over keys in given [`Schema`] in all snapshots in reverse lexicographical order
+pub struct SnapshotManagerIter<'a, S: Schema> {
+    db_iter: Peekable<RawDbReverseIterator<'a>>,
+    snapshot_iterators: Vec<Peekable<SchemaBatchIterator<'a, S>>>,
+}
+
+impl<'a, S: Schema> SnapshotManagerIter<'a, S> {
+    fn new(
+        db_iter: RawDbReverseIterator<'a>,
+        snapshot_iterators: Vec<SchemaBatchIterator<'a, S>>,
+    ) -> Self {
+        Self {
+            db_iter: db_iter.peekable(),
+            snapshot_iterators: snapshot_iterators
+                .into_iter()
+                .map(|iter| iter.peekable())
+                .collect(),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum DataLocation {
+    Db,
+    // Index inside `snapshot_iterators`
+    Snapshot(usize),
+}
+
+impl<'a, S: Schema> Iterator for SnapshotManagerIter<'a, S> {
+    type Item = (SchemaKey, SchemaValue);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Find max value
+        loop {
+            let mut max_values: Vec<(DataLocation, &SchemaKey)> = vec![];
+            let max_db_value = self.db_iter.peek();
+            if let Some((db_key, _)) = max_db_value {
+                max_values.push((DataLocation::Db, db_key));
+            };
+
+            for (idx, iter) in self.snapshot_iterators.iter_mut().enumerate() {
+                if let Some(&(peeked_key, _)) = iter.peek() {
+                    if max_values.is_empty() {
+                        max_values.push((DataLocation::Snapshot(idx), peeked_key));
+                    } else {
+                        let (_, max_key) = &max_values[0];
+                        match peeked_key.cmp(max_key) {
+                            Ordering::Greater => {
+                                max_values.clear();
+                                max_values.push((DataLocation::Snapshot(idx), peeked_key));
+                            }
+                            Ordering::Equal => {
+                                max_values.push((DataLocation::Snapshot(idx), peeked_key));
+                            }
+                            Ordering::Less => {}
+                        }
+                    }
+                }
+            }
+
+            if max_values.is_empty() {
+                break;
+            }
+
+            // We don't need key anymore
+            let mut max_values: Vec<DataLocation> = max_values
+                .into_iter()
+                .map(|(location, _)| location)
+                .collect();
+
+            // Save location of max value to be probably returned
+            let last_max_location = max_values.pop().unwrap();
```

Comment:
It looks like we convert a vec into iter (line 155) thne back into a vec just to get firsy element? Woul `max_values[0].location` work?

Reply:
We need to do conversion, so it can drop &SchemaKey it got from iterator, only to keep index.

If we do just by index, then we have life time issue. Let's consider this change:

```rust
// // We don't need key anymore
            // let mut max_values: Vec<DataLocation> = max_values
            //     .into_iter()
            //     .map(|(location, _)| location)
            //     .collect();

            // Save location of max value to be probably returned
            let (last_max_location, _) = max_values.last().unwrap();

            // Move all iterators to next value
            for (location, _) in max_values.iter() {
                match location {
                    DataLocation::Db => {
                        let _ = self.db_iter.next().unwrap();
                    }
                    Snapshot(idx) => {
                        let _ = self.snapshot_iterators[*idx].next().unwrap();
                    }
                }
            }
```

We will get this error:

```
   --> full-node/sov-prover-storage-manager/src/snapshot_manager.rs:167:33
    |
125 |             let max_db_value = self.db_iter.peek();
    |                                ------------ first mutable borrow occurs here
...
164 |             for (location, _) in max_values.iter() {
    |                                  ----------------- first borrow later used here
...
167 |                         let _ = self.db_iter.next().unwrap();
    |                                 ^^^^^^^^^^^^ second mutable borrow occurs here

```

So we use original max_values vector. 

Also, we need "last" element, not first.

Reply:
Fixed